src/v/raft/mutex_buffer.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+#include "raft/types.h"
+#include "ssx/future-util.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/shared_ptr.hh>
+
+namespace raft::details {
+
+template<typename Request, typename Response>
+class mutex_buffer {
+public:
+    explicit mutex_buffer(mutex& m, size_t max_buffered_elements)
+      : _mutex(m)
+      , _max_buffered(max_buffered_elements) {}
+
+    ss::future<Response> enqueue(Request&& r) {
+        if (_gate.is_closed()) {
+            return ss::make_exception_future<Response>(
```

Comment:
is this a private op? why would enqueue _expose_ a gate closed exception? seems like implementation detail to me. 

What is the actual error ?

Reply:
the error is that the gate was closed, we are not longer able to handle the request, I can do `ss::with_gate`.